scripts/publish.ts

```diff
@@ -0,0 +1,408 @@
+            } else {
+              return [packageName, packageVersionOrReleaseType] as const;
+            }
+          },
+        ),
+      ),
+    );
+  } else {
+    // We're creating an unstable release, so for each package, we need to:
+    // 1. Browse the version history to find the highest published prerelease
+    //    number that matches the current branch / prerelease identifier.
+    // 2. Then:
+    //    - If we find a matching prerelease number, increment it and keep the
+    //      main part of the SemVer string (i.e., the `<major>.<minor>.<patch>`
+    //      part) unchanged.
+    //    - If we don't find any, initialize the prerelease number to `1` and
+    //      increment the patch version number instead.
+    const packageNameToTaggedVersion = new Map(
+      [...(await readCurrentBranchHeadTags())]
+        .map(parsePackageVersionSpecifier)
+        .filter(isDefined),
+    );
+    const prereleaseIdentifier = isOnDefaultBranch
+      ? maturityLevel
+      : `${maturityLevel}.${sanitizedCurrentBranchName}`;
+    packageNameToReleaseVersion = new Map(
+      await Promise.all(
+        packageNames.map(async (packageName) => {
+          const taggedVersion = packageNameToTaggedVersion.get(packageName);
+          if (taggedVersion !== undefined) {
+            // There's no need to increment the prerelease number if the HEAD
+            // already has a tagged version.
+            return [packageName, taggedVersion] as const;
+          }
+          const packageVersion = await readPackageVersion(packageName);
+          const parsedPackageVersion = semverParse(packageVersion);
+          if (parsedPackageVersion === null) {
+            throw new CommandError(
+              formatInvalidPackageVersionError(packageName, packageVersion),
+            );
+          }
+          const highestPublishedPrereleaseNumber =
+            findHighestPublishedPrereleaseNumber(
+              [...(packageNameToVersionHistory.get(packageName) ?? [])],
+              parsedPackageVersion,
+              prereleaseIdentifier,
+            );
+          const incrementedPackageVersion = incrementVersionByPrerelease(
+            parsedPackageVersion,
+            prereleaseIdentifier,
+            highestPublishedPrereleaseNumber,
+          );
+          return [packageName, incrementedPackageVersion] as const;
+        }),
+      ),
+    );
+  }
+  function formatInvalidPackageVersionError(
+    packageName: string,
+    packageVersion: string,
+  ): string {
+    return `Package \`${packageName}\` has an invalid version \`${packageVersion}\` in its manifest.`;
+  }
+
+  // Ensure the new versions aren't already in use.
+  if (!args["tolerate-republish"]) {
+    packageNameToReleaseVersion.forEach(
+      (packageReleaseVersion, packageName) => {
+        const packageVersionHistory =
+          packageNameToVersionHistory.get(packageName);
+        if (packageVersionHistory?.has(packageReleaseVersion)) {
```

Comment:
Hum it's more because yarn already do it that for us, it's not required in our script. It's good now that we have it, but I feel it's an edge case since yarn handle it for us properly.